### ISSUE TITLE
Poblem with $(location :foo) macro expansion in cross-cell context

### ISSUE THREAD

**Issue by davido** [facebook/buck#544]
I'm trying to isolate third party library in its own cell inside my project:

```
repositories.jgit = lib/jgit
```

The `jgit_src` is defined as:

```
$ buck targets --json //lib/jgit:jgit_src
[
{
  "binaryJar" : ":jgit__download_src",
  "buck.base_path" : "",
  "buck.direct_dependencies" : [ "//lib:LICENSE-jgit", "//lib/jgit:jgit__download_src" ],
  "buck.type" : "prebuilt_jar",
  "deps" : [ "//lib:LICENSE-jgit" ],
  "gwtJar" : null,
  "javadocUrl" : null,
  "mavenCoords" : null,
  "name" : "jgit_src",
  "sourceJar" : null,
  "visibility" : [ "PUBLIC" ]
}
]
```

And the target output is located:

```
$ buck targets --show_output //lib/jgit:jgit_src
/lib/jgit:jgit_src buck-out/gen/lib/jgit/jgit_src.jar

$ ls -all buck-out/gen/lib/jgit/jgit_src.jar
-rw-r--r-- 1 davido users 2025387 Dec  4 00:23 buck-out/gen/lib/jgit/jgit_src.jar
```

Now, the consumer is saying this:

```
genrule(
  name = 'jgit_edit_src',
  cmd = 'unzip -qd $TMP $(location @jgit//:jgit_src) ' +
    'org/eclipse/jgit/diff/Edit.java;' +
    'cd $TMP;' +
    'zip -Dq $OUT org/eclipse/jgit/diff/Edit.java',
  out = 'edit.src.zip',
)
```

But the `location` macros expansion is relocated in root context of cell container:

```
(cd /home/davido/projects/gerrit/buck-out/gen/gerrit-patch-jgit/jgit_edit_src__srcs && OUT=/home/davido/projects/gerrit/buck-out/gen/gerrit-patch-jgit/jgit_edit_src/edit.src.zip TMP=/home/davido/projects/gerrit/buck-out/gen/gerrit-patch-jgit/jgit_edit_src__tmp /bin/bash -e -c 'unzip -qd $TMP /home/davido/projects/gerrit/buck-out/gen/jgit_src.jar org/eclipse/jgit/diff/Edit.java;cd $TMP;zip -Dq $OUT org/eclipse/jgit/diff/Edit.java')
```

I would expect, that it should expand to `buck-out/gen/lib/jgit/jgit_src.jar` instead.

**Comment by sdwilsh** [comment]
Keep in mind that cross-cell could point to something completely outside of the tree, so I don't think we'd ever put out `buck-out/gen/lib/jgit/jgit_src.jar` as the path to the output.

**Comment by davido** [comment]
Probably related to this is similar problem in `$(classpath)` macro expansion:

```
cmd.extend(['--lib', '$(classpath %s)' % '//gerrit-war:init'])
```

renders result relocated in root context of cell container:

`/home/davido/projects/gerrit/buck-out/gen/jgit.jar`

This file doesn't exist. The correct location is: `/home/davido/projects/gerrit/lib/jgit/buck-out/gen/jgit.jar`.

Interestingly, when querying the same from `buck audit classpath` command, the result is correct:

```
$ buck audit classpath //gerrit-war:init | grep -i buck-out/gen/jgit.jar
/home/davido/projects/gerrit/lib/jgit/buck-out/gen/jgit.jar
```

Amd not relocated to the root directory of cell container.

**Comment by davido** [comment]
> Keep in mind that cross-cell could point to something completely outside of the tree, so I don't think we'd ever put out buck-out/gen/lib/jgit/jgit_src.jar as the path to the output.

Yes, my mistake. It should probably expand to:

```
/home/davido/projects/gerrit/lib/jgit/buck-out/gen/jgit_src.jar
```

In any event, this file should be built and accessible from whatever `$(location @jgit//:jgit_src)` is expand to.

**Comment by davido** [comment]
Closed with 53cdfbc232a6890a13c0d94c40f13e9449c10529.
